hre/src/main/java/hre/util/Notifier.scala

```diff
@@ -0,0 +1,95 @@
+package hre.util
+
+import java.io.{ByteArrayInputStream, File}
+
+import sys.process._
+import hre.lang.System.Warning
+
+import scala.collection.JavaConverters._
+import java.nio.file.{Files, Paths}
+
+object Notifier {
+  def notify(title: String, message: String): Boolean = {
+    val os = System.getProperty("os.name").toLowerCase()
+    if (os.contains("win")) {
+      notifyWindows10(title, message)
+    } else if (os.contains("mac")) {
+      notifyMacOS(title, message)
+    } else if (os.contains("nix") || os.contains("linux")) {
+      notifyLibnotify(title, message)
+    } else {
+      false
+    }
+  }
+
+  def notifyLibnotify(title: String, message: String): Boolean = {
+    if (commandExists("notify-send")) {
+      val cmd = Seq("notify-send", title, message)
+      cmd ! ProcessLogger(_ => Unit, _ => Unit) match {
+        case 0 => true
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  def notifyMacOS(title: String, message: String): Boolean = {
+    if (commandExists("osascript")) {
+      val cmd = Seq("osascript", "-e", s"""display notification "$message" with title "$title"""")
+      cmd ! ProcessLogger(_ => Unit, _ => Unit) match {
+        case 0 => true
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  val powershellNotificationScript =
+    """
+      |$app = '{1AC14E77-02E7-4E5D-B744-2EB1AE5198B7}\WindowsPowerShell\v1.0\powershell.exe'
+      |[Windows.UI.Notifications.ToastNotificationManager, Windows.UI.Notifications, ContentType = WindowsRuntime]
+      |
+      |$Template = [Windows.UI.Notifications.ToastTemplateType]::ToastImageAndText01
+      |
+      |#Gets the Template XML so we can manipulate the values
+      |[xml]$ToastTemplate = ([Windows.UI.Notifications.ToastNotificationManager]::GetTemplateContent($Template).GetXml())
+      |
+      |[xml]$ToastTemplate = @"
+      |<toast launch="app-defined-string">
+      |  <visual>
+      |    <binding template="ToastGeneric">
+      |      <text>%s</text>
+      |      <text>%s</text>
+      |    </binding>
+      |  </visual>
+      |</toast>
+      |"@
+      |
+      |$ToastXml = New-Object -TypeName Windows.Data.Xml.Dom.XmlDocument
+      |$ToastXml.LoadXml($ToastTemplate.OuterXml)
+      |
+      |$notify = [Windows.UI.Notifications.ToastNotificationManager]::CreateToastNotifier($app)
+      |
+      |$notify.Show($ToastXml)
+      |""".stripMargin
+
+  def notifyWindows10(title: String, message: String): Boolean = {
+    // Script comes from: https://gist.github.com/Windos/9aa6a684ac583e0d38a8fa68196bc2dc
+    val script = powershellNotificationScript.format(title, message)
```

Comment:
this seems suspect, surely you need some escape here?

Reply:
You mean, escaping "title" and "message" for use in powershell, or escaping the powershellNotificationScript variable? Assuming you mean the second case: VerCors doesn't really have a thread model/security concerns afaik, but if we choose to be defensive I'll add an escaping step.

Reply:
I mean the first case, surely if your message is ">" or something it would break.

Reply:
I guess I'll just whitelist "a-zA-Z!? ", that's easy enough and very effective.

Reply:
Oh, that is also fine (did not see your reply in time)